nncf/common/graph/patterns.py

```diff
@@ -0,0 +1,199 @@
+                add_subgraph_to_graph(final_graph, subgraph_copy)
+
+        final_pattern = copy.deepcopy(self)
+        final_pattern._graph = final_graph
+        return final_pattern
+
+    def __or__(self, other: 'GraphPattern') -> 'GraphPattern':
+        """
+        Add other's DiGraph nodes to self's DiGraph as a new weakly connected components.
+        It is a syntax sugar of 'add_pattern_alternative()'
+
+        :param other: GraphPattern that will be added
+        """
+        new_pattern = copy.deepcopy(self)
+        other_copy = create_copy_of_subgraph(other.graph)
+        add_subgraph_to_graph(new_pattern._graph, other_copy)
+        return new_pattern
+
+    def __eq__(self, other: 'GraphPattern') -> bool:
+        return ism.is_isomorphic(self.graph, other.graph)
+
+    @property
+    def graph(self) -> nx.DiGraph:
+        return self._graph
+
+    def add_pattern_alternative(self, other: 'GraphPattern') -> None:
+        """
+        Add 'other' pattern as a connected weakly component to 'self' pattern.
+
+        :param other: GraphPattern that will be added
+        """
+        other_graph = other.graph
+        self_graph = self.graph
+        for node in other_graph.nodes:
+            if node in self_graph.nodes:
+                assert False
+            self_graph.add_node(node, type=other_graph.nodes[node]['type'], label=other_graph.nodes[node]['label'])
+        for edge in other_graph.edges:
+            self_graph.add_edge(edge[0], edge[1])
+
+    def join_patterns(self, other: 'GraphPattern',
+                      edges: Optional[List[Tuple[Hashable, Hashable]]] = None) -> None:
+        """
+        Add 'other' pattern to 'self' pattern and connect nodes from self to other determined by 'edges'.
+        If edges is None, add edge between
+        last node of self's graph and first node of other's graph.
+
+        The first and last nodes are found by nx.lexicographical_topological_sort().
+
+        :param other: GraphPattern that will be added
+        :param edges: List of edges between self and other graphs.
+            Edges must begin at self and finish at other.
+        """
+        if edges is None:
+            first_node_other = list(nx.lexicographical_topological_sort(other.graph, key=int))[0]
+            assert other.graph.in_degree(first_node_other) == 0
+            last_node_self = list(nx.lexicographical_topological_sort(self.graph, key=int))[-1]
+            assert self.graph.out_degree(last_node_self) == 0
+            edges = [[last_node_self, first_node_other]]
+        for edge in edges:
+            assert edge[0] in self.graph.nodes
+            assert edge[1] in other.graph.nodes
+        add_subgraph_to_graph(self.graph, other.graph)
+        for edge in edges:
+            self.graph.add_edge(edge[0], edge[1])
+
+    def add_node(self, label: str, node_type: Union[str, List[str]]) -> int:
```

Comment:
Node type should go first, label should go second and be optional

Reply:
Update to more arbitary variant -> `add_node(self, **attrs) -> int:`